prow/tide/search.go

```diff
@@ -0,0 +1,147 @@
+	timeFormatISO8601 = "2006-01-02T15:04:05Z"
+)
+
+type searchExecutor func(start, end time.Time) ([]PullRequest, int /*true match count*/, error)
+
+func newSearchExecutor(ctx context.Context, ghc githubClient, log *logrus.Entry, q string) searchExecutor {
+	return func(start, end time.Time) ([]PullRequest, int, error) {
+		q = fmt.Sprintf("%s %s", q, dateToken(start, end))
+		vars := map[string]interface{}{
+			"query":        githubql.String(q),
+			"searchCursor": (*githubql.String)(nil),
+		}
+		var totalCost, remaining int
+		var totalMatches int
+		var ret []PullRequest
+		for {
+			sq := searchQuery{}
+			if err := ghc.Query(ctx, &sq, vars); err != nil {
+				return nil, 0, err
+			}
+			totalCost += int(sq.RateLimit.Cost)
+			remaining = int(sq.RateLimit.Remaining)
+			totalMatches = int(sq.Search.IssueCount)
+			// If the search won't return all results, abort.
+			if totalMatches > 1000 {
+				return nil, totalMatches, nil
+			}
+			for _, n := range sq.Search.Nodes {
+				ret = append(ret, n.PullRequest)
+			}
+			if !sq.Search.PageInfo.HasNextPage {
+				break
+			}
+			vars["searchCursor"] = githubql.NewString(sq.Search.PageInfo.EndCursor)
+		}
+		log.Debugf("Search for query \"%s\" returned %d PRs and cost %d point(s). %d remaining.", q, len(ret), totalCost, remaining)
+		return ret, totalMatches, nil
+	}
+}
+
+func (q searchExecutor) search() ([]PullRequest, error) {
+	prs, _, err := q.searchRange(time.Time{}, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchSince(t time.Time) ([]PullRequest, error) {
+	prs, _, err := q.searchRange(t, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchRange(start, end time.Time) ([]PullRequest, int, error) {
+	prs, count, err := q(start, end)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if count <= 1000 {
+		// The search returned all the results for the query.
+		return prs, len(prs), nil
+	}
+	// The query returned too many results, we need to partition it.
+	prs, err = q.partitionSearchRange(start, end, count)
+	return prs, len(prs), err
+}
+
+func (q searchExecutor) partitionSearchRange(start, end time.Time, count int) ([]PullRequest, error) {
+	partition := partitionTime(start, end, count, 900)
+	// Search right side...
+	rPRs, rCount, err := q.searchRange(partition, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Search left side...
+	// For the left side we can deduce the count in advance.
+	lCount := count - rCount
+	// If the count is too large we can skip the initial search and go straight to
+	// partitioning to save an API token.
+	var lPRs []PullRequest
+	if lCount <= 1000 {
+		lPRs, _, err = q.searchRange(start, partition)
+	} else {
+		lPRs, err = q.partitionSearchRange(start, partition, lCount)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return append(lPRs, rPRs...), nil
+}
+
+func partitionTime(start, end time.Time, count, goalSize int) time.Time {
+	// TODO(cjwagner): Special case old start times. (Nobody has PRs from before 2000)
```

Comment:
To clarify: these dates are GitHub side set on PR open, NOT git dates, correct?
If that's the case, then this definitely works since GitHub was founded in 2007, but it might be helpful to more explicitly note _why_ 2000 works somewhere.

Reply:
Yes these are the last updated times of the PRs.